crates/host/src/wasmbus/mod.rs

```diff
@@ -305,6 +306,280 @@ impl Handler {
+                },
+            )
+            .await?;
+        let wasmcloud_compat::blobstore::GetObjectResponse {
+            success,
+            error,
+            initial_chunk,
+            ..
+        } = decode_provider_response(res)?;
+        match (success, initial_chunk, error) {
+            (_, _, Some(err)) => Err(anyhow!(err).context("failed to get object response")),
+            (false, _, None) => bail!("failed to get object response"),
+            (true, None, None) => Ok((Box::new(empty()), 0)),
+            (
+                true,
+                Some(wasmcloud_compat::blobstore::Chunk {
+                    object_id,
+                    container_id,
+                    bytes,
+                    ..
+                }),
+                None,
+            ) => {
+                ensure!(object_id == name);
+                ensure!(container_id == container);
+                let size = bytes
+                    .len()
+                    .try_into()
+                    .context("value size does not fit in `u64`")?;
+                Ok((Box::new(Cursor::new(bytes)), size))
+            }
+        }
+    }
+
+    #[instrument]
+    async fn has_object(&self, container: &str, name: String) -> anyhow::Result<bool> {
+        let targets = self.targets.read().await;
+        self.call_operation(
+            targets.get(&TargetInterface::WasiBlobstoreBlobstore),
+            "wasmcloud:blobstore/Blobstore.ObjectExists",
+            &wasmcloud_compat::blobstore::ContainerObject {
+                container_id: container.into(),
+                object_id: name,
+            },
+        )
+        .await
+        .and_then(decode_provider_response)
+    }
+
+    #[instrument(skip(value))]
+    async fn write_data(
+        &self,
+        container: &str,
+        name: String,
+        mut value: Box<dyn AsyncRead + Sync + Send + Unpin>,
+    ) -> anyhow::Result<()> {
+        let targets = self.targets.read().await;
+        let mut bytes = Vec::new();
+        value
+            .read_to_end(&mut bytes)
+            .await
+            .context("failed to read bytes")?;
+        let res = self
+            .call_operation(
+                targets.get(&TargetInterface::WasiBlobstoreBlobstore),
+                "wasmcloud:blobstore/Blobstore.PutObject",
+                &wasmcloud_compat::blobstore::PutObjectRequest {
+                    chunk: wasmcloud_compat::blobstore::Chunk {
+                        object_id: name,
+                        container_id: container.into(),
+                        bytes,
+                        offset: 0,
+                        is_last: true,
+                    },
+                    ..Default::default()
+                },
+            )
+            .await?;
+        let wasmcloud_compat::blobstore::PutObjectResponse { stream_id } =
+            decode_provider_response(res)?;
+        ensure!(stream_id.is_none());
```

Comment:
Nit: add a string here explaining that multipart uploads aren't supported(?)

Reply:
Do you mean a comment?
Because this is not user-facing (or at least it should not be)

Reply:
Do providers return a `stream_id` even when the complete blob is sent?

Reply:
looks like it should not https://github.com/wasmCloud/wasmCloud/blob/a82847f4e4a2637bc9d6f9c75be0a902f7944212/crates/providers/blobstore-fs/src/bin/blobstore_fs.rs#L590-L599